tensorflow/compiler/xla/tests/tanh_test.cc

```diff
@@ -0,0 +1,149 @@
+#include <cmath>
+#include <limits>
+#include <memory>
+#include <numeric>
+#include <random>
+#include <vector>
+
+#include "absl/base/casts.h"
+#include "absl/types/span.h"
+#include "tensorflow/compiler/xla/array2d.h"
+#include "tensorflow/compiler/xla/array3d.h"
+#include "tensorflow/compiler/xla/array4d.h"
+#include "tensorflow/compiler/xla/client/global_data.h"
+#include "tensorflow/compiler/xla/client/local_client.h"
+#include "tensorflow/compiler/xla/client/xla_builder.h"
+#include "tensorflow/compiler/xla/layout_util.h"
+#include "tensorflow/compiler/xla/literal.h"
+#include "tensorflow/compiler/xla/statusor.h"
+#include "tensorflow/compiler/xla/test.h"
+#include "tensorflow/compiler/xla/tests/client_library_test_base.h"
+#include "tensorflow/compiler/xla/tests/literal_test_util.h"
+#include "tensorflow/compiler/xla/tests/test_macros.h"
+#include "tensorflow/compiler/xla/types.h"
+#include "tensorflow/core/platform/types.h"
+
+namespace xla {
+namespace {
+
+class TanhTest : public ClientLibraryTestBase {};
+
+XLA_TEST_F(TanhTest, TanhF32sNormalTest) {
+  // This is like the test TanhTest.TanhF32s above, except that
+  // the input tensor is large enough to exercise the vectorized tanh
+  // implementation on XLA CPU.
+  XlaBuilder builder(TestName());
+  auto input_literal = LiteralUtil::CreateR1<float>(
+      {1.02,  -0.32, 0.85,  0.90,  1.23,  -0.91, -0.49, 0.80,  -0.67, 0.16,
```

Comment:
We actually have exhaustive tanh f32 tests: https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/compiler/xla/tests/exhaustive_unary_test_f32_or_smaller.cc;l=453-463?q=tensorflow%2Fcompiler%2Fxla%2Ftests%2Fexhaustive_unary_test_f32_or_smaller.cc

could we tighten the error spec there and add required invariants instead?

Reply:
@x@ 
I don't know much about XLA, I don't think my single test is suitable for embedding ```Exhaustive32BitOrLessUnaryTest```.
1. ```Exhaustive32BitOrLessUnaryTest``` is a batch test, but bounder test is not need batch test.
2.  input of ```Exhaustive32BitOrLessUnaryTest``` must be int type, but I need a float or double input .

Reply:
the unittest of tanh is not very suitable to be placed in this test.
1. the interface of ```run``` only recv a xla_op, I need ```Max(Tanh(input), input_boundary);``` to express ```output < 1```
2. I want to express ```output < 1```, but error_spec_gen only express ```expect - 0.0001 < output < expect + 0.0001```

Reply:
Would it be possible to modify that test instead then? E.g. modify the `Run` method to accept an optional `check_invariant` callback, which will check the desired property (|output| < 1 in your case)? That seems better than having a hardcoded set of outputs.

Reply:
done